src/graphql_clj/spec.clj

```diff
@@ -205,64 +266,42 @@
 
 (defmethod spec-for :default [_ _])
 
-(defmulti ^:private of-type (fn [n _] (:kind n)))
-
-(defmethod ^:private of-type :LIST [{:keys [inner-type]} s]
-  (loop [it inner-type]
-    (if (:type-name it)
-      (named-spec s [(:type-name it)])
-      (recur (:inner-type it)))))
-
-(defmethod ^:private of-type :default [{:keys [spec]} _]
-  spec)
-
-;; Parent and base types
-
-(defn get-type-node [spec s]
-  "Given a spec, get the corresponding node from the AST"
-  (get-in s [:spec-map spec]))
-
-(defn get-base-type-node
-  "Given a spec, get the node definition for the corresponding base type"
-  [{:keys [spec]} s]
-  (when-let [base-spec (s/get-spec spec)]
-    (if (default-type-names (name base-spec))
-      {:node-type :scalar :type-name (name base-spec)}
-      (get-type-node base-spec s))))
-
-(defn get-parent-type
-  "Given a node and the global state, find the parent type"
-  [{:keys [v/parent]} s]
-  (if-let [base-parent (get-type-node (of-type parent s) s)]
-    (of-type base-parent s)
-    (recur parent s)))
+(defn- safe-eval [recursive? d]
+  (if (or recursive? (not (meta d)))
+    (do
+      (assert (= (first d) 'clojure.spec/def))               ;; Protect against unexpected statement eval
+      (try (eval d) (catch Compiler$CompilerException _ d))) ;; Squashing errors here to provide better error messages in validation
```

Comment:
`(assert  (= (first d) 'clojure.spec/def))` does not prevent against all unexpected statement eval. The best way is not to use eval at all.